File: 0129-sum-root-to-leaf-numbers/0129-sum-root-to-leaf-numbers.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sumNumbers(self, root: Optional[TreeNode]) -> int:

        def dfs(root, acc):
            if not root.left and not root.right:
                acc.append(str(root.val))
                tmp = [int(''.join(acc))]
                acc.pop()
                return tmp
            
            acc.append(str(root.val))
            answer = []
            if root.left:
                answer += dfs(root.left, acc)
            if root.right:
                answer += dfs(root.right, acc)
            acc.pop()

            return answer
        
        return sum(dfs(root, []))
        

File: 0129-sum-root-to-leaf-numbers/0129-sum-root-to-leaf-numbers.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int sumNumbers(TreeNode root) {
        return dfs(root, new ArrayList<>()).stream().mapToInt(Integer::valueOf).sum();
    }

    private List<Integer> dfs(TreeNode root, List<String> digits) {
        digits.add(String.valueOf(root.val)) ;

        if (root.left == null && root.right == null) {
            var num = toInt(digits);
            digits.remove(digits.size() - 1);
            return List.of(num);
        }

        var answer = new ArrayList<Integer>();
        if (root.left != null) {
            answer.addAll(dfs(root.left, digits));
        }
        if (root.right != null) {
            answer.addAll(dfs(root.right, digits));
        }

        digits.remove(digits.size() - 1);
        return answer;
    }

    private int toInt(List<String> digits) {
        return Integer.parseInt(
                String.join("", digits)
        );
    }
}